src/popsift/common/excl_blk_prefix_sum.h

```diff
@@ -73,7 +73,6 @@ class Block
         if( threadIdx.x == 0 && threadIdx.y == 0 ) {
             loop_total = 0;
         }
-        __syncthreads();
```

Comment:
> Pre-RTX cards have no per-thread program counter and can fall out of lockstep if syncthreads is missing after a thread-conditional if-than-else.

This doesn't match my understanding of what `__syncthreads()` does going back at least as far as compute capability 3.0. All threads in a warp execute in lock-step, and branch divergence is handled using an activity mask. In the example below, my understanding is that the `__syncthreads()` is *not* required after branch divergence:

```c++
// ... do work ...
if(threadIdx.x == 1) {
  // Diverge and do something only on T1.
  printf("Hello from T1\n");
}
__syncthreads(); // Not required
// ... do work ...
```

My understanding is that the only constrain on `__syncthreads()` is that it must be hit by all threads in the same block. For example, the following block is incorrect because not all threads are waiting on the same instruction:

```c++
if(threadIdx.x == 1) __syncthreads();
__syncthreads();
```

Here are a few of the resources I'm looking at:
- [C Programming Guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#synchronization-functions)
- [Demystifying GPU Microarchitecture through Microbenchmarking](www.stuffedcow.net/files/gpuarch-ispass2010.pdf) - see section D on pages 5-6.

Any thoughts?

Reply:
We started using CUDA when it was CC 1.3, so a lot of stuff that I have gotten used to may have been out-dated for several generations.

You are absolutely correct that nothing bad happens without those syncthreads after the end of these ifs. With RTXes, it is 100% certain that there is no penalty.

But I'm not sure if it a great idea to the syncthreads after a divergent if: The most recent measurements that show that reconvergence fails are in an arXiv paper from 2015: https://arxiv.org/pdf/1504.01650.pdf
They are observing a measurable speed degradation up to CC 5 when they store content in a global memory array after a loop that diverges.

Reply:
Thank you for the context. I'll reintroduce those syncthreads() if there's no penalty.